

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tiger Lily 0.1-dev documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="None" href="#" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li><a href="#">Tiger Lily 0.1-dev documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="toctree-wrapper compound">
<span id="document-srcdocs/modules"></span><div class="section" id="project-modules">
<h1>Project Modules<a class="headerlink" href="#project-modules" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<span id="document-srcdocs/tigerlily"></span><div class="section" id="tigerlily-package">
<h2>tigerlily Package<a class="headerlink" href="#tigerlily-package" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3><tt class="xref py py-mod docutils literal"><span class="pre">tigerlily</span></tt> Package<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<span class="target" id="module-tigerlily.__init__"></span></div>
<div class="section" id="subpackages">
<h3>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
<span id="document-srcdocs/tigerlily.grc"></span><div class="section" id="grc-package">
<h4>grc Package<a class="headerlink" href="#grc-package" title="Permalink to this headline">¶</a></h4>
<div class="section" id="id1">
<h5><tt class="xref py py-mod docutils literal"><span class="pre">grc</span></tt> Package<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h5>
<span class="target" id="module-tigerlily.grc"></span><p>tigerlily.grc - functions and extensions for common GRC-related tasks.</p>
<p>The members of this package generally interact with online or downloaded
resources from the members of the GRC (Genome Reference Consortium) and
other online resource entities such as NCBI, Sanger Institute, and the
UCSC Genome Browser.</p>
</div>
<div class="section" id="module-tigerlily.grc.genome">
<span id="genome-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">genome</span></tt> Module<a class="headerlink" href="#module-tigerlily.grc.genome" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.grc.genome.GRCGenome">
<em class="property">class </em><tt class="descclassname">tigerlily.grc.genome.</tt><tt class="descname">GRCGenome</tt><a class="reference internal" href="_modules/tigerlily/grc/genome.html#GRCGenome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.GRCGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.grc.genome.ReferenceGenome" title="tigerlily.grc.genome.ReferenceGenome"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.grc.genome.ReferenceGenome</span></tt></a></p>
<p>Fetch, store, load, parse, and extract sequences from a GCR ref assembly</p>
<p>NCBI has released many top-quality reference genomes. As of late 2010, these
assemblies are produced under the brand of the GRC - the Genome Reference
Consortium. The UCSC Genome Browser mirrors each of these assemblies in
such a way that makes it very conveniant to download and extract data that
is useful to Tiger Lily.</p>
<p>This class provides an interface to automatically download, store, load,
parse, and extract sequences from the UCSC Genome Browser&#8217;s copies of the
GRC reference genomes.</p>
<p>Support for different assemblies will be added manually to this class. For
a list of supported assemblies by their name, see SUPPORTED_ASSEMBLIES .
The default (most current) assembly will be stored in DEFAULT_ASSEMBLY</p>
<p>For the convenience of faster non-networked tests, extremely small made-up
reference genomes are provided inside of this package in a folder called
&#8216;test_assemblies&#8217;. They can be loaded either by using the GRCGenome.load
method (as normal), or else by using GRCGenome.download with names that
start with &#8216;test&#8217; (eg. &#8216;test1&#8217;, &#8216;test2&#8217;, &#8216;testnomask&#8217;, etc.).</p>
<dl class="classmethod">
<dt id="tigerlily.grc.genome.GRCGenome.download">
<em class="property">classmethod </em><tt class="descname">download</tt><big>(</big><em>name='h19'</em>, <em>store=False</em>, <em>silent=True</em><big>)</big><a class="reference internal" href="_modules/tigerlily/grc/genome.html#GRCGenome.download"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.GRCGenome.download" title="Permalink to this definition">¶</a></dt>
<dd><p>Download a reference genome of the given name, and return a GRCGenome</p>
<p>Fetches the named reference assembly (default is DEFAULT_ASSEMBLY) from
the web, and creates a new GRCGenome object to handle it.</p>
<p>If store is False (default), the data will be kept entirely in memory,
and will be destroyed as soon as the object is released. If True,
the .tar.gz of the entire assembly will be saved in the current
directory - ValueError will be raised if this file seems to already
exist. If store is a string, it will be assumed to be a path to a
directory (trailing slash optional) in which the .tar.gz archive should
be stored. (Again, ValueError will be raised if the file already
exists.)</p>
<p>If silent is False, status messages will be printed using print() to 
keep the user informed of the progress. This is usually very important
in command line applications as the reference archives are about 900 MB
in size and may take minutes or hours to download depending on the
internet connection.</p>
<p>Because of the large size of these files, it is highly recommended that
the store option be set. Please do not use Tiger Lily to abuse the
UCSC Genome Browser group&#8217;s generosity in hosting these large files to
the general public.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">refgen</span> <span class="o">=</span> <span class="n">GRCGenome</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s">&#39;test1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">refgen2</span> <span class="o">=</span> <span class="n">GRCGenome</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s">&#39;test1&#39;</span><span class="p">,</span><span class="n">store</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s">&#39;test1.tar.gz&#39;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="s">&#39;test1.tar.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Only supported reference genome assemblies are allowed, otherwise
ValueError will be raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">GRCGenome</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s">&#39;invalid&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Unknown or unsupported reference genome specified</span>
</pre></div>
</div>
</dd></dl>

<dl class="classmethod">
<dt id="tigerlily.grc.genome.GRCGenome.load">
<em class="property">classmethod </em><tt class="descname">load</tt><big>(</big><em>path</em><big>)</big><a class="reference internal" href="_modules/tigerlily/grc/genome.html#GRCGenome.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.GRCGenome.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Load the given .tar.gz archive file as a downloaded GRC Genome.</p>
<p>It is expected that this file will be named &lt;assembly&gt;.tar.gz</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">refgen</span> <span class="o">=</span> <span class="n">GRCGenome</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s">&#39;test1&#39;</span><span class="p">,</span><span class="n">store</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s">&#39;test1.tar.gz&#39;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">refgen2</span> <span class="o">=</span> <span class="n">GRCGenome</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s">&#39;test1.tar.gz&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">refgen2</span><span class="o">.</span><span class="n">sequences</span><span class="p">())</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">refgen</span><span class="o">.</span><span class="n">sequences</span><span class="p">())</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="s">&#39;test1.tar.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="classmethod">
<dt id="tigerlily.grc.genome.GRCGenome.load_archive">
<em class="property">classmethod </em><tt class="descname">load_archive</tt><big>(</big><em>archive</em><big>)</big><a class="reference internal" href="_modules/tigerlily/grc/genome.html#GRCGenome.load_archive"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.GRCGenome.load_archive" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an instance of tarfile.TarFile, load it as a ref. assembly.</p>
</dd></dl>

<dl class="method">
<dt id="tigerlily.grc.genome.GRCGenome.sequences">
<tt class="descname">sequences</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/grc/genome.html#GRCGenome.sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.GRCGenome.sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a MixedSequenceGroup of the FASTA sequences from this ref.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">PolymerSequenceGroup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">refgen</span> <span class="o">=</span> <span class="n">GRCGenome</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s">&#39;test1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="n">refgen</span><span class="o">.</span><span class="n">sequences</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span><span class="n">PolymerSequenceGroup</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.grc.genome.ReferenceGenome">
<em class="property">class </em><tt class="descclassname">tigerlily.grc.genome.</tt><tt class="descname">ReferenceGenome</tt><a class="reference internal" href="_modules/tigerlily/grc/genome.html#ReferenceGenome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.ReferenceGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Abstract base class for all Genome objects.</p>
<p>ReferenceGenome objects represent an underlying file or data structure that
encodes a Reference Genome assembly (often called simply an &#8216;assembly&#8217;,
although in Tiger Lily they will normally use the full name Reference Genome
to help clarity).</p>
<p>ReferenceGenome objects convert such file structures in to sequence groups,
normally of the type tigerlily.sequences.MixedGroup, with the contained
sequences being tigerlily.sequences.NucleicSequence - one per each
assembly unit (nominally chromosomes, but often with other meta-units 
included), with the identifier set appropriately.</p>
<dl class="method">
<dt id="tigerlily.grc.genome.ReferenceGenome.sequences">
<tt class="descname">sequences</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/grc/genome.html#ReferenceGenome.sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.grc.genome.ReferenceGenome.sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a PolymerSequenceGroup representing this Genome.</p>
<p>Normally this will be a MixedSequenceGroup with NucleicSequence members.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<span id="document-srcdocs/tigerlily.index"></span><div class="section" id="index-package">
<h4>index Package<a class="headerlink" href="#index-package" title="Permalink to this headline">¶</a></h4>
<div class="section" id="module-tigerlily.index.fixedtree">
<span id="fixedtree-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">fixedtree</span></tt> Module<a class="headerlink" href="#module-tigerlily.index.fixedtree" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.index.fixedtree.FixedTree">
<em class="property">class </em><tt class="descclassname">tigerlily.index.fixedtree.</tt><tt class="descname">FixedTree</tt><big>(</big><em>sequence_group</em>, <em>width</em>, <em>reverse=False</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTree" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.index.index.GroupIndex" title="tigerlily.index.index.GroupIndex"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.index.index.GroupIndex</span></tt></a></p>
<p>GroupIndex wrapper which implements a Fixed-Width Substring Tree.</p>
<p>The resulting tree supports alignments of a fixed width only. It supports
memory and time efficient lookups including mismatches. It does not support
gaps, indels, arbitrary width lookups, etc.</p>
<p>As an example, the following block of code will create a MixedSequenceGroup
that we can index.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">NucleicSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">createNucleicSequenceGroup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;ACGTACTTAGCATCATACGTCAGTACGCAGTCAGTCAGTCAT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;CGAGCGACGCAGTACGTACTGGCAGACGTGTATACCTGC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">group</span> <span class="o">=</span> <span class="n">createNucleicSequenceGroup</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span> <span class="o">=</span> <span class="n">FixedTree</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="tigerlily.index.fixedtree.FixedTree.add_sequence">
<tt class="descname">add_sequence</tt><big>(</big><em>sequence</em>, <em>reverse</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTree.add_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTree.add_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the given sequence to this index.</p>
<p>The sequence must be a NucleicSequence, and it will be split in to
subsequences of the length specified by this index.</p>
<p>If reverse is True, each individual subsequence will also be reversed.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">NucleicSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">createNucleicSequenceGroup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;ACGTACTTAGCATCATACGTCAGTACGCAGTCAGTCAGTCAT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;CGAGCGACGCAGTACGTACTGGCAGACGTGTATACCTGC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">group</span> <span class="o">=</span> <span class="n">createNucleicSequenceGroup</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span><span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span> <span class="o">=</span> <span class="n">FixedTree</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;CCCCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span><span class="o">.</span><span class="n">add_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="tigerlily.index.fixedtree.FixedTree.alignments">
<tt class="descname">alignments</tt><big>(</big><em>sequence</em>, <em>mismatches=0</em>, <em>maximum_alignments=None</em>, <em>best_alignments=False</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTree.alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTree.alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of all alignments produced by the given input.</p>
<dl class="docutils">
<dt>Retrurns a list of tuples, each tuple containing three values. They are:</dt>
<dd><p class="first">0: str - &#8216;chromosome name&#8217;
1: int - &#8216;position&#8217;
2: boolean - &#8216;strand&#8217; (True means &#8216;reported strand&#8217;, False means</p>
<blockquote class="last">
<div>&#8216;opposing strand&#8217;. Opposing strand matches
do not alter the original position.)</div></blockquote>
</dd>
</dl>
<p>Each reported alignment will have a computed &#8216;Hamming Distance&#8217; (see
<a class="reference external" href="http://en.wikipedia.org/wiki/Hamming_distance">http://en.wikipedia.org/wiki/Hamming_distance</a> ) of no greater than the
mismatches argument. If left at 0, no actual edit distance calculations
are performed.</p>
<p>If maximum_alignments is an integer greater than 0, then only that many
alignments will be found - after that many are found, the search ends.</p>
<p>If best_alignments is False, then every possible alignment will be found
even if maximum_alignments is set. Then, alignments are reported in
increasing order of their edit distance from the search sequence. In
general, this function will slow down the search considerably,
particularly if maximum_alignments is set (since the primary benefit of
maximim_alignments will be negated.)</p>
<p>This will raise ValueError if the given sequence does not match the
pre-specified width of the index.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">NucleicSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">createNucleicSequenceGroup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;GGAATTCC&#39;</span><span class="p">,</span><span class="n">identifier</span><span class="o">=</span><span class="s">&#39;foo&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqgroup</span> <span class="o">=</span> <span class="n">createNucleicSequenceGroup</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span> <span class="o">=</span> <span class="n">FixedTree</span><span class="p">(</span><span class="n">seqgroup</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index</span><span class="o">.</span><span class="n">alignments</span><span class="p">(</span><span class="s">&#39;GG&#39;</span><span class="p">)</span>
<span class="go">[(&#39;foo&#39;, 0, True)]</span>
</pre></div>
</div>
<p>Note that for the next example with mismatches, the order of the result
is unspecified, so we will just wrap it with len() to avoid testing
errors. You don&#8217;t need to wrap the result in len() in your own code.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">index</span><span class="o">.</span><span class="n">alignments</span><span class="p">(</span><span class="s">&#39;GG&#39;</span><span class="p">,</span><span class="n">mismatches</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>For the next example we enable best_alignments, so the order is 
gaurunteed.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">index</span><span class="o">.</span><span class="n">alignments</span><span class="p">(</span><span class="s">&#39;GG&#39;</span><span class="p">,</span><span class="n">mismatches</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">best_alignments</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="go">[(&#39;foo&#39;, 0, True), (&#39;foo&#39;, 1, True)]</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.index.fixedtree.FixedTreeNode">
<em class="property">class </em><tt class="descclassname">tigerlily.index.fixedtree.</tt><tt class="descname">FixedTreeNode</tt><big>(</big><em>alignment=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTreeNode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTreeNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Implicit tree data structure for storing a fixed read length index.</p>
<p>Each node in the graph may contain alignments (although the Root node is
gaurunteed to not store any alignments). Any input sequence that lands on
a node with a set of alignments may report those alignments.</p>
<p>Each node in the graph may contain a dictionary that maps strings to other
nodes - these strings represent labeled edges. Moving along an edge consumes
the corresponding prefix from the input sequence.</p>
<p>Hamming distance may be used to move along an edge that doesn&#8217;t exactly
match to a prefix of the input sequence. If the Hamming distance between an
edge and a prefix of the input sequence is less than or equal to the number
of remaining mismatches, then travel may proceed across that edge, although
the mismatch count is decremented by that hamming distance.</p>
<p>Great care is made to ensure that the following properties are maintained in
this structure at all times:</p>
<blockquote>
<div><ul class="simple">
<li>For any given node N, for all of N&#8217;s edges E1, there exists no edge
in N called E2 that is not E1 but for which E2 is an exact prefix
of E1.</li>
<li>For any given node N, for all of N&#8217;s children M, there is no way to
re-distribute the edges between N and the M&#8217;s (even by creating new
children nodes in-between) in order to increase the length of an
edge&#8217;s label without violating the first property.</li>
</ul>
</div></blockquote>
<p>In other words, the edges of each node are &#8216;maximally uncommon&#8217; - they are
chosen to be as long as possible without sharing any prefixes.</p>
<p>Because of this property, we can be sure of correctness and optimality.</p>
<p>Well, I&#8217;m mostly hoping about the optimality part. I haven&#8217;t done the math.</p>
<dl class="method">
<dt id="tigerlily.index.fixedtree.FixedTreeNode.alignments">
<tt class="descname">alignments</tt><big>(</big><em>sequence</em>, <em>original_mismatches</em>, <em>remaining_mismatches=0</em>, <em>maximum_alignments=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTreeNode.alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTreeNode.alignments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="tigerlily.index.fixedtree.FixedTreeNode.insert">
<tt class="descname">insert</tt><big>(</big><em>sequence</em>, <em>alignment</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/fixedtree.html#FixedTreeNode.insert"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.fixedtree.FixedTreeNode.insert" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-tigerlily.index.index">
<span id="index-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">index</span></tt> Module<a class="headerlink" href="#module-tigerlily.index.index" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.index.index.GroupIndex">
<em class="property">class </em><tt class="descclassname">tigerlily.index.index.</tt><tt class="descname">GroupIndex</tt><big>(</big><em>sequence_group</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/index.html#GroupIndex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.index.GroupIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Abstract base class for all genomic indexes.</p>
<dl class="method">
<dt id="tigerlily.index.index.GroupIndex.alignments">
<tt class="descname">alignments</tt><big>(</big><em>sequence</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/tigerlily/index/index.html#GroupIndex.alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.index.index.GroupIndex.alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Gather all alignments for the given sequence. (See __contains__).</p>
<p>Implementing subclasses may (and almost certainly will) provide
additional arguments to constrain the alignment set, such as to
allow for a certain number of mismatches or to only generate the
closest fit, etc.</p>
<p>The return value will always be a list or list-like object, with
each item corresponding to an alignment. (The list may be empty.)
However, each alignment&#8217;s representation is <em>undefind</em>. This is
because different indexing methods may or may not be able to
provide the different details allowed by other indexing methods.
Consult the implementing subclasses&#8217; documentation for the structure
of an alignment.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<span id="document-srcdocs/tigerlily.sequences"></span><div class="section" id="sequences-package">
<h4>sequences Package<a class="headerlink" href="#sequences-package" title="Permalink to this headline">¶</a></h4>
<div class="section" id="id1">
<h5><tt class="xref py py-mod docutils literal"><span class="pre">sequences</span></tt> Package<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h5>
<span class="target" id="module-tigerlily.sequences"></span></div>
<div class="section" id="module-tigerlily.sequences.fasta">
<span id="fasta-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">fasta</span></tt> Module<a class="headerlink" href="#module-tigerlily.sequences.fasta" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.sequences.fasta.FASTA">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.fasta.</tt><tt class="descname">FASTA</tt><big>(</big><em>file=None</em>, <em>data=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTA" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequenceGroup" title="tigerlily.sequences.sequence.PolymerSequenceGroup"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequenceGroup</span></tt></a></p>
<p>Implementation of an NCBI-compatible FASTA format.</p>
<dl class="classmethod">
<dt id="tigerlily.sequences.fasta.FASTA.load_sequences">
<em class="property">classmethod </em><tt class="descname">load_sequences</tt><big>(</big><em>*sequences</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTA.load_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTA.load_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new FASTA object from a list of arbitrary sequences.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences</span> <span class="kn">import</span> <span class="n">RawSequence</span><span class="p">,</span><span class="n">NucleicSequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">RawSequence</span><span class="p">(</span><span class="s">&#39;AACGGTTACGATCAGGACTACGGGAGGAGAGA&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;ACGGACTTACCAGGACTACGGACTCAGACG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta</span> <span class="o">=</span> <span class="n">FASTA</span><span class="o">.</span><span class="n">load_sequences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">fasta</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.fasta.FASTA.write">
<tt class="descname">write</tt><big>(</big><em>file</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTA.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTA.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write this FASTA group to a file, producing a conforming FASTA file.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="s">r&#39;&#39;&#39;&gt;SEQUENCE_1</span>
<span class="gp">... </span><span class="s">MTEITAAMVKELRESTGAGMMDCKNALSETNGDFDKAVQLLREKGLGKAAKKADRLAAEG</span>
<span class="gp">... </span><span class="s">LVSVKVSDDFTIAAMRPSYLSYEDLDMTFVENEYKALVAELEKENEERRRLKDPNKPEHK</span>
<span class="gp">... </span><span class="s">IPQFASRKQLSDAILKEAEEKIKEELKAQGKPEKIWDNIIPGKMNSFIADNSQLDSKLTL</span>
<span class="gp">... </span><span class="s">MGQFYVMDDKKTVEQVIAEKEKEFGGKIKIVEFICFEVGEGLEKKTEDFAAEVAAQL</span>
<span class="gp">... </span><span class="s">&gt;SEQUENCE_2</span>
<span class="gp">... </span><span class="s">SATVSEINSETDFVAKNDQFIALTKDTTAHIQSNSLQSVEELHSSTINGVKFEEYLKSQI</span>
<span class="gp">... </span><span class="s">ATIGENLVVRRFATLKAGANGVVNGYIHTNGRVGVVIAAACDSAEVASKSRDLLRQICMH</span>
<span class="gp">... </span><span class="s">&#39;&#39;&#39;</span>
</pre></div>
</div>
<p>We could use the file-based loading initializer for FASTA, but let&#8217;s
just leave it as a string for simplicity.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="n">FASTA</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>Now we write the sequences out to a buffer.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">io</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">buffer</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">buffer</span><span class="p">)</span>
</pre></div>
</div>
<p>You&#8217;ll just have to trust that the output is consistent, and that
it only differs from data in terms of superficial formatting.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.sequences.fasta.FASTASequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.fasta.</tt><tt class="descname">FASTASequence</tt><big>(</big><em>sequence</em>, <em>identifier</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTASequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTASequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.FormattedSequence" title="tigerlily.sequences.sequence.FormattedSequence"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.FormattedSequence</span></tt></a></p>
<p>Container for a single FASTA Sequence.</p>
<p>Do not use this object. Instead, use the FASTA object, which subclasses
PolymerSequenceGroup. The FASTA format is intrinsically a set, so most of
the parsing logic is left in the FASTA object.</p>
<p>Nothing is <em>stopping</em> you from using this object, but you will probably
find that it doesn&#8217;t do very much work for you.</p>
<dl class="attribute">
<dt id="tigerlily.sequences.fasta.FASTASequence.MAX_LINE_WIDTH">
<tt class="descname">MAX_LINE_WIDTH</tt><em class="property"> = 79</em><a class="headerlink" href="#tigerlily.sequences.fasta.FASTASequence.MAX_LINE_WIDTH" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.fasta.FASTASequence.identifier">
<tt class="descname">identifier</tt><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTASequence.identifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTASequence.identifier" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.fasta.FASTASequence.sequence">
<tt class="descname">sequence</tt><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTASequence.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTASequence.sequence" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.fasta.FASTASequence.write">
<tt class="descname">write</tt><big>(</big><em>file</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/fasta.html#FASTASequence.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.fasta.FASTASequence.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write this FASTA sequence to the opened file object.</p>
<p>Note that this function has the same result as writing the output of
.format(). However, this function will outperform* that approach for
large sequences (such as reference genomes stored in FASTA format),
because this function doesn&#8217;t store the entire sequence in memory a
second time like .format() does (for text wrapping purposes).</p>
<dl class="docutils">
<dt><a href="#id2"><span class="problematic" id="id3">*</span></a>: OK, the performance will be mostly identical, it&#8217;s just that the</dt>
<dd>memory footprint should be much smaller.</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-tigerlily.sequences.genomic">
<span id="genomic-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">genomic</span></tt> Module<a class="headerlink" href="#module-tigerlily.sequences.genomic" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.sequences.genomic.AminoSequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.genomic.</tt><tt class="descname">AminoSequence</tt><big>(</big><em>sequence</em>, <em>identifier=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#AminoSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.AminoSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequence" title="tigerlily.sequences.sequence.PolymerSequence"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequence</span></tt></a></p>
<p>PolymerSequence for aminoacid sequences (e.g. protein sequences).</p>
<p>Each member of the sequence must be one of ABCDEFGHIKLMNOPQRSTUVWYZX*</p>
<dl class="attribute">
<dt id="tigerlily.sequences.genomic.AminoSequence.identifier">
<tt class="descname">identifier</tt><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#AminoSequence.identifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.AminoSequence.identifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the identifier, if any.</p>
<p>If the identifier has not been explicitly set, the default identifier
will be used instead.</p>
</dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.genomic.AminoSequence.sequence">
<tt class="descname">sequence</tt><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#AminoSequence.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.AminoSequence.sequence" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.genomic.AminoSequence.translations">
<tt class="descname">translations</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#AminoSequence.translations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.AminoSequence.translations" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate every NucleicSequence that could translate to this.</p>
<p>Because any given AminoSequence might have multiple nucleic acid
sequences that could translate in to it, this generator function will
iterate over every possible NucleicSequence object that could make this
AminoSequence.</p>
<p>Note that the generated NucleicSequence objects do not use any sort
of genomic signaling or encoding parameters - in other words, stop and
start codons are not implicitly added unless the AminoSequence 
contained them to begin with.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">AminoSequence</span><span class="p">(</span><span class="s">&#39;NDC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trans</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">sequence</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">s1</span><span class="o">.</span><span class="n">translations</span><span class="p">()]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">trans</span><span class="p">)</span>
<span class="go">8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s">&#39;AATGATTGT&#39;</span> <span class="ow">in</span> <span class="n">trans</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.sequences.genomic.NucleicSequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.genomic.</tt><tt class="descname">NucleicSequence</tt><big>(</big><em>sequence</em>, <em>identifier=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequence" title="tigerlily.sequences.sequence.PolymerSequence"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequence</span></tt></a></p>
<p>Container for nucleic (chromosomal DNA) sequences.</p>
<p>In general this is not an &#8216;input&#8217; or &#8216;output&#8217; format, but rather an
internal format that constrains the sequence to a certain set of
allowed characters in the sequence.</p>
<dl class="docutils">
<dt>The set of characters allowed is the following:</dt>
<dd>ATGC</dd>
</dl>
<p>If a sequence is converted to a NucleicSequence and doesn&#8217;t fit that
set, ValueError will be raised.</p>
<p>NucleicSequence objects do not have a valid write() method, nor do
they have a valid format() method. (Instead they will raise
NotImplementedError.)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tigerlily.sequences.raw</span> <span class="kn">as</span> <span class="nn">raw</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqdata</span> <span class="o">=</span> <span class="s">&#39;CTAGCATACTCACAGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">raw</span><span class="o">.</span><span class="n">RawSequence</span><span class="p">(</span><span class="n">seqdata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleic</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="n">NucleicSequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleic</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;CTAGCATACTCACAGT&#39;</span>
</pre></div>
</div>
<p>An example showing that NucleicSequence objects can be instantiated
directly.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="n">seqdata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">True</span>
</pre></div>
</div>
<p>An example where the constraint fails:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">raw</span><span class="o">.</span><span class="n">RawSequence</span><span class="p">(</span><span class="s">&#39;CAGTTACTm&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleic</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="n">NucleicSequence</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError: Invalid nucleic sequence format</span>: <span class="n">CAGTTACTm</span>
</pre></div>
</div>
<dl class="method">
<dt id="tigerlily.sequences.genomic.NucleicSequence.complement">
<tt class="descname">complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the purine&lt;-&gt;pyrimidine complement of the sequence
as a new sequence.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;AATGCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cseq1</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cseq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;TTACGG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">cseq1</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">seq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.genomic.NucleicSequence.identifier">
<tt class="descname">identifier</tt><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.identifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.identifier" title="Permalink to this definition">¶</a></dt>
<dd><p>The identifier for this read.</p>
</dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.genomic.NucleicSequence.reverse">
<tt class="descname">reverse</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.reverse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new sequence that is the reverse of this sequence.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;AATGCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rseq1</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rseq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;CCGTAA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">rseq1</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">seq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.genomic.NucleicSequence.reverse_complement">
<tt class="descname">reverse_complement</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse complement of the sequence as a new sequence.</p>
<p>Has the same result as sequence.reverse().complement(), but slightly
more efficient.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;AATGCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rcseq1</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rcseq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;GGCATT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">rcseq1</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">seq1</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rcseq1</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">seq1</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.genomic.NucleicSequence.sequence">
<tt class="descname">sequence</tt><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.sequence" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.genomic.NucleicSequence.translate">
<tt class="descname">translate</tt><big>(</big><em>reading_frame=1</em>, <em>use_control_codes=False</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#NucleicSequence.translate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.NucleicSequence.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this NucleicSequence to a corresponding AminoSequence.</p>
<p>Conversion is performed by taking each codon and replacing it with
the corresponding amino acid. In biological terms,
translate() assumes that the NucleicSequence object&#8217;s sequence is
the coding sequence (and not the template). Incidentally, to get
the coding sequence of a template, just call
template.reverse_complement().</p>
<p>reading_frame is either 1, 2, or 3, and corresponds to the 1-based
offset into the sequence from which to begin collecting codons. In
other words, reading_frame=1 (the default) will start from the first
base in the sequence, and reading_frame=2 from the 2nd, and so on.
Higher values are allowed and will start further in to the sequence by
skipping codons that would otherwise have been in the reading frame.</p>
<p>use_control_codes is a flag which, when True, enables an alternative
processing algorithm. The new algorithm is like the first one (including
the reading_frame), but after processing has finished the following
constraint is placed upon the strand: all amino acids prior to and
including the first Methionine (M/ATG) will be removed, and all amino
acids after the first STOP codon (<a href="#id4"><span class="problematic" id="id5">*</span></a>/TAA,TGA,TAG) will be removed. If
no Methionine is detected, ValueError will be raised. (It is not
necessary for there to be a STOP codon.)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">nucleic</span> <span class="o">=</span> <span class="n">NucleicSequence</span><span class="p">(</span><span class="s">&#39;CATGGTATGTTTTGGGTTTAGAAACGT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span> <span class="o">=</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;HGMFWV*KR&#39;</span>
</pre></div>
</div>
<p>We can also specify a reading_frame even though it will mean that the
given input sequence doesn&#8217;t cleanly subdivide in to codons - this is
not an error.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span> <span class="o">=</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">reading_frame</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;MVCFGFRN&#39;</span>
</pre></div>
</div>
<p>Here is an example using use_control_codes mode.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span> <span class="o">=</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">use_control_codes</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;FWV&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span> <span class="o">=</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">use_control_codes</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span><span class="n">reading_frame</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span><span class="o">.</span><span class="n">sequence</span>
<span class="go">&#39;VCFGFRN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">amino</span> <span class="o">=</span> <span class="n">nucleic</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">use_control_codes</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span><span class="n">reading_frame</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">No Methionine found in translated nucleic sequence</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="tigerlily.sequences.genomic.createNucleicSequenceGroup">
<tt class="descclassname">tigerlily.sequences.genomic.</tt><tt class="descname">createNucleicSequenceGroup</tt><big>(</big><em>*sequences</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#createNucleicSequenceGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.createNucleicSequenceGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert any group of sequences in to a nucleic MixedSequence group.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tigerlily.sequences.raw</span> <span class="kn">import</span> <span class="n">Raw</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="n">Raw</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s">&#39;AGTACGTATTTCAT</span><span class="se">\n</span><span class="s">TTCATACGACTAC</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleic</span> <span class="o">=</span> <span class="n">createNucleicSequenceGroup</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">nucleic</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">nucleic</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span><span class="n">NucleicSequence</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">True</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="tigerlily.sequences.genomic.reverse_complement">
<tt class="descclassname">tigerlily.sequences.genomic.</tt><tt class="descname">reverse_complement</tt><big>(</big><em>sequence</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/genomic.html#reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.genomic.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the reverse complement of the input string.</p>
<p>Input is assumed to be a string that could be the sequence of a
NucleicSequence.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement</span><span class="p">(</span><span class="s">&#39;ACGGTC&#39;</span><span class="p">)</span>
<span class="go">&#39;GACCGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement</span><span class="p">(</span><span class="s">&#39;GACCGT&#39;</span><span class="p">)</span>
<span class="go">&#39;ACGGTC&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-tigerlily.sequences.mixed">
<span id="mixed-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">mixed</span></tt> Module<a class="headerlink" href="#module-tigerlily.sequences.mixed" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.sequences.mixed.MixedSequenceGroup">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.mixed.</tt><tt class="descname">MixedSequenceGroup</tt><big>(</big><em>sequence_group=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/mixed.html#MixedSequenceGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.mixed.MixedSequenceGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequenceGroup" title="tigerlily.sequences.sequence.PolymerSequenceGroup"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequenceGroup</span></tt></a></p>
<p>PolymerSequenceGroup that allows sequences of any type.</p>
<p>What this gains in flexibility, it loses in representation. This
subclass of PolymerSequenceGroup does not have a .write() method, since
there probably isn&#8217;t a good way to represent the sequences it contains.</p>
<p>We also gain an add() method, which allows additional sequences to be
added to the sequence group after initialization - perhaps the key
benefit of a MixedSequenceGroup.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tigerlily.sequences</span> <span class="kn">as</span> <span class="nn">tigseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">tigseq</span><span class="o">.</span><span class="n">FASTASequence</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s">&#39;aCGTAtagcATCA&#39;</span><span class="p">,</span><span class="n">identifier</span><span class="o">=</span><span class="s">&#39;seq1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">tigseq</span><span class="o">.</span><span class="n">RawSequence</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s">&#39;GGCATACGGCAatacgaCATN&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="n">tigseq</span><span class="o">.</span><span class="n">Raw</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="s">&#39;GGCATACT</span><span class="se">\n</span><span class="s">GAGcgaACT</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomic</span> <span class="o">=</span> <span class="n">tigseq</span><span class="o">.</span><span class="n">MixedSequenceGroup</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">genomic</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomic</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomic</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">genomic</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomic</span> <span class="o">=</span> <span class="n">tigseq</span><span class="o">.</span><span class="n">MixedSequenceGroup</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">genomic</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genomic</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">genomic</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
<dl class="method">
<dt id="tigerlily.sequences.mixed.MixedSequenceGroup.add">
<tt class="descname">add</tt><big>(</big><em>sequence</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/mixed.html#MixedSequenceGroup.add"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.mixed.MixedSequenceGroup.add" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-tigerlily.sequences.raw">
<span id="raw-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">raw</span></tt> Module<a class="headerlink" href="#module-tigerlily.sequences.raw" title="Permalink to this headline">¶</a></h5>
<dl class="class">
<dt id="tigerlily.sequences.raw.Raw">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.raw.</tt><tt class="descname">Raw</tt><big>(</big><em>file=None</em>, <em>data=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#Raw"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.Raw" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequenceGroup" title="tigerlily.sequences.sequence.PolymerSequenceGroup"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequenceGroup</span></tt></a></p>
<p>Container for a set of raw (unformatted) sequences, one per line.</p>
<p>Use this object to parse data that stores sequences one per line without
any additional formatting or information beyond the read and a newline
character for each read (except possibly the last one).</p>
<p>Additionally, blank lines are skipped without an error (even if they 
contain white space characters).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="s">r&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="s">CGTATACGCTCAGTC</span>
<span class="gp">... </span><span class="s">CGGGGCATCAGACTA</span>
<span class="gp">... </span><span class="s">CACGTACGACTACGTACGACTGACTGACTGCATCACATG</span>
<span class="gp">... </span><span class="s"></span>
<span class="gp">... </span><span class="s">LAGVVGALVUIALKT</span>
<span class="gp">... </span><span class="s">&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="n">Raw</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
<span class="go">4</span>
</pre></div>
</div>
<dl class="method">
<dt id="tigerlily.sequences.raw.Raw.write">
<tt class="descname">write</tt><big>(</big><em>file</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#Raw.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.Raw.write" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.sequences.raw.RawSequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.raw.</tt><tt class="descname">RawSequence</tt><big>(</big><em>sequence</em>, <em>identifier=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#RawSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.RawSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.FormattedSequence" title="tigerlily.sequences.sequence.FormattedSequence"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.FormattedSequence</span></tt></a></p>
<p>Container for a &#8216;raw&#8217; sequence.</p>
<p>Since raw reads are essentially just a sequence, this is a fairly useless
class. It is more or less just a holder for the Raw sequence group object.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">RawSequence</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s">&#39;ATCGCGAGTCAGTCAGCATGACTACGCACAGTAC&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>One possible use of this container is to create Raw sequences that know
their identifier. Raw sequences that are given an identifier don&#8217;t forget
it and will happily send it when being converted to and from different
formats. They just don&#8217;t ever use them in output.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">RawSequence</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s">&#39;LALL&#39;</span><span class="p">,</span><span class="n">identifier</span><span class="o">=</span><span class="s">&#39;seq1&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>An example converting from Raw to FASTA and back, to prove it works:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tigerlily.sequences</span> <span class="kn">as</span> <span class="nn">sequence</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasta</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">FASTASequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="n">RawSequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">fasta</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="n">seq2</span><span class="o">.</span><span class="n">sequence</span> <span class="o">==</span> <span class="s">&#39;LALL&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span><span class="o">.</span><span class="n">identifier</span> <span class="o">==</span> <span class="n">fasta</span><span class="o">.</span><span class="n">identifier</span> <span class="o">==</span> <span class="n">seq2</span><span class="o">.</span><span class="n">identifier</span> <span class="o">==</span> <span class="s">&#39;seq1&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="tigerlily.sequences.raw.RawSequence.identifier">
<tt class="descname">identifier</tt><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#RawSequence.identifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.RawSequence.identifier" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.raw.RawSequence.sequence">
<tt class="descname">sequence</tt><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#RawSequence.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.RawSequence.sequence" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.raw.RawSequence.write">
<tt class="descname">write</tt><big>(</big><em>file</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/raw.html#RawSequence.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.raw.RawSequence.write" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-tigerlily.sequences.sequence">
<span id="sequence-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">sequence</span></tt> Module<a class="headerlink" href="#module-tigerlily.sequences.sequence" title="Permalink to this headline">¶</a></h5>
<p>Module providing an abstraction of nucleic or amino acid polymer sequences.</p>
<dl class="class">
<dt id="tigerlily.sequences.sequence.FormattedSequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.sequence.</tt><tt class="descname">FormattedSequence</tt><big>(</big><em>sequence=None</em>, <em>identifier=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#FormattedSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.FormattedSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="index.html#tigerlily.sequences.sequence.PolymerSequence" title="tigerlily.sequences.sequence.PolymerSequence"><tt class="xref py py-class docutils literal"><span class="pre">tigerlily.sequences.sequence.PolymerSequence</span></tt></a></p>
<p>Abstract base class for PolymerSequence objects which can be formatted.</p>
<p>Children of this class are sequences with some sort of character-encoded
format. In particular, they support a  .format() method and a .write()
method.</p>
<dl class="method">
<dt id="tigerlily.sequences.sequence.FormattedSequence.format">
<tt class="descname">format</tt><big>(</big><em>to_type=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#FormattedSequence.format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.FormattedSequence.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a string representing this sequence in the perscribed format.</p>
<p>The last character of the format is gaurunteed to be a newline.</p>
<p>To return this string in the format that the Sequence is already in,
you may omit to_type.</p>
<p>Note that some FormattedSequence descendants will be representing a
binary format. These descendants can choose whether they wish to return
a &#8216;bytes&#8217; object instead of a string, or simply raise an exception.</p>
</dd></dl>

<dl class="method">
<dt id="tigerlily.sequences.sequence.FormattedSequence.write">
<tt class="descname">write</tt><big>(</big><em>file</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#FormattedSequence.write"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.FormattedSequence.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the formatted output of this sequence in to the file.</p>
<p>For some formats this function may not make sense (such as for
tile-and-cycle based outputs like Illumina&#8217;s BCL format), and
may return NotImplementedError instead.</p>
<dl class="docutils">
<dt>The result, otherwise, will be the same as if you had called:</dt>
<dd>file.write(sequence.format())</dd>
</dl>
<p>However, note that for some formats, calling this method is
considerably faster than using .format(), particularly for large
sequences.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.sequences.sequence.PolymerSequence">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.sequence.</tt><tt class="descname">PolymerSequence</tt><big>(</big><em>sequence=None</em>, <em>identifier=None</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#PolymerSequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.PolymerSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">builtins.object</span></tt></p>
<p>Abstract base class for representing genomic sequences.</p>
<p>See the documentation for this module for examples on using this class.</p>
<dl class="method">
<dt id="tigerlily.sequences.sequence.PolymerSequence.convert">
<tt class="descname">convert</tt><big>(</big><em>to_type</em><big>)</big><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#PolymerSequence.convert"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.PolymerSequence.convert" title="Permalink to this definition">¶</a></dt>
<dd><p>Return this sequence as an instance of to_type.</p>
<p>This can be useful in changing a sequence from one format to another.
For this purpose, see format()</p>
<p>No checking is made to ensure that to_type is a subclass of
PolymerSequence, but it would probably be bad to pass in a type that
isn&#8217;t a PolymerSequence type.</p>
</dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.sequence.PolymerSequence.identifier">
<tt class="descname">identifier</tt><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#PolymerSequence.identifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.PolymerSequence.identifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a str object identifying the sequence.</p>
<p>Ideally this will come straight from the underlying source data (e.g.
the header row on FASTA sequences), however, some data sources have
no given identifier. In this case it is better to give SOME sort of
contextualizing identifier than to give up and use a placeholder like
&#8216;Unknown&#8217;. An example would be to return the line number the sequence
occured on, or the file that the sequence was contained in, or the
date and time the sequence was processed, etc.</p>
<p>If all else fails and you can&#8217;t identify the sequence properly, just
use &#8216;return super()&#8217;. Currently this will return the string &#8216;Unknown&#8217;.</p>
</dd></dl>

<dl class="attribute">
<dt id="tigerlily.sequences.sequence.PolymerSequence.sequence">
<tt class="descname">sequence</tt><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#PolymerSequence.sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.PolymerSequence.sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a str object representing the sequence.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="tigerlily.sequences.sequence.PolymerSequenceGroup">
<em class="property">class </em><tt class="descclassname">tigerlily.sequences.sequence.</tt><tt class="descname">PolymerSequenceGroup</tt><a class="reference internal" href="_modules/tigerlily/sequences/sequence.html#PolymerSequenceGroup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.sequences.sequence.PolymerSequenceGroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">_abcoll.Iterable</span></tt></p>
<p>Abstract base class for representing groups of genomic sequences.</p>
<p>This could be used for any purpose you like, but was originally intended
for use when the underlying format implies a grouping of PolymerSequence
objects, such as in the FASTA format.</p>
</dd></dl>

</div>
</div>
<span id="document-srcdocs/tigerlily.utility"></span><div class="section" id="utility-package">
<h4>utility Package<a class="headerlink" href="#utility-package" title="Permalink to this headline">¶</a></h4>
<div class="section" id="id1">
<h5><tt class="xref py py-mod docutils literal"><span class="pre">utility</span></tt> Package<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h5>
<span class="target" id="module-tigerlily.utility"></span></div>
<div class="section" id="module-tigerlily.utility.string_relations">
<span id="string-relations-module"></span><h5><tt class="xref py py-mod docutils literal"><span class="pre">string_relations</span></tt> Module<a class="headerlink" href="#module-tigerlily.utility.string_relations" title="Permalink to this headline">¶</a></h5>
<dl class="function">
<dt id="tigerlily.utility.string_relations.greatest_common_prefix">
<tt class="descclassname">tigerlily.utility.string_relations.</tt><tt class="descname">greatest_common_prefix</tt><big>(</big><em>s1</em>, <em>s2</em><big>)</big><a class="reference internal" href="_modules/tigerlily/utility/string_relations.html#greatest_common_prefix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.utility.string_relations.greatest_common_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the longest common prefix between s1 and s2.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">greatest_common_prefix</span><span class="p">(</span><span class="s">&#39;banana&#39;</span><span class="p">,</span><span class="s">&#39;bandit&#39;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">greatest_common_prefix</span><span class="p">(</span><span class="s">&#39;apple&#39;</span><span class="p">,</span><span class="s">&#39;sour apple&#39;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">greatest_common_prefix</span><span class="p">(</span><span class="s">&#39;tree&#39;</span><span class="p">,</span><span class="s">&#39;tree&#39;</span><span class="p">)</span>
<span class="go">4</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="tigerlily.utility.string_relations.hamming_distance">
<tt class="descclassname">tigerlily.utility.string_relations.</tt><tt class="descname">hamming_distance</tt><big>(</big><em>s1</em>, <em>s2</em><big>)</big><a class="reference internal" href="_modules/tigerlily/utility/string_relations.html#hamming_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.utility.string_relations.hamming_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Hamming edit distance between s1 and s2.</p>
<p>The Hamming edit distance is defined as the number of individual alterations
performed on characters in one string in order to turn it in to another
string of the same length.</p>
<p>If s1 and s2 are not the same length, ValueError will be raised.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">hamming_distance</span><span class="p">(</span><span class="s">&#39;party&#39;</span><span class="p">,</span><span class="s">&#39;party&#39;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hamming_distance</span><span class="p">(</span><span class="s">&#39;zebra&#39;</span><span class="p">,</span><span class="s">&#39;cobra&#39;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hamming_distance</span><span class="p">(</span><span class="s">&#39;one&#39;</span><span class="p">,</span><span class="s">&#39;three&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Cannot compute Hamming distance of strings of unequal length</span>
</pre></div>
</div>
<p>With apologies to the fine editors of wikipedia.com for kernel of this
code.</p>
</dd></dl>

<dl class="function">
<dt id="tigerlily.utility.string_relations.levenshtein_distance">
<tt class="descclassname">tigerlily.utility.string_relations.</tt><tt class="descname">levenshtein_distance</tt><big>(</big><em>s1</em>, <em>s2</em><big>)</big><a class="reference internal" href="_modules/tigerlily/utility/string_relations.html#levenshtein_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#tigerlily.utility.string_relations.levenshtein_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Levenshtein edit distance (integer) between s1 and s2.</p>
<p>The Levenshtein is defined as the minimum number of substitutions,
additions, and deletions needed to transform s1 in to s2.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">levenshtein_distance</span><span class="p">(</span><span class="s">&#39;kitten&#39;</span><span class="p">,</span><span class="s">&#39;sitten&#39;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">levenshtein_distance</span><span class="p">(</span><span class="s">&#39;sitten&#39;</span><span class="p">,</span><span class="s">&#39;sittin&#39;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">levenshtein_distance</span><span class="p">(</span><span class="s">&#39;sittin&#39;</span><span class="p">,</span><span class="s">&#39;sitting&#39;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">levenshtein_distance</span><span class="p">(</span><span class="s">&#39;Alabama&#39;</span><span class="p">,</span><span class="s">&#39;Hell&#39;</span><span class="p">)</span> <span class="c"># Surprisingly, not 0!</span>
<span class="go">7</span>
</pre></div>
</div>
<p>Thanks to hetland.org for this code:
hetland.org/coding/python/levenshtein.py
(no copyright information was found)</p>
</dd></dl>

</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-srcdocs/modules">Project Modules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-srcdocs/tigerlily">tigerlily Package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#id1"><tt class="docutils literal"><span class="pre">tigerlily</span></tt> Package</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#subpackages">Subpackages</a><ul>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-srcdocs/tigerlily.grc">grc Package</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#id1"><tt class="docutils literal"><span class="pre">grc</span></tt> Package</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.grc.genome"><tt class="docutils literal"><span class="pre">genome</span></tt> Module</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-srcdocs/tigerlily.index">index Package</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.index.fixedtree"><tt class="docutils literal"><span class="pre">fixedtree</span></tt> Module</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.index.index"><tt class="docutils literal"><span class="pre">index</span></tt> Module</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-srcdocs/tigerlily.sequences">sequences Package</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#id1"><tt class="docutils literal"><span class="pre">sequences</span></tt> Package</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.sequences.fasta"><tt class="docutils literal"><span class="pre">fasta</span></tt> Module</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.sequences.genomic"><tt class="docutils literal"><span class="pre">genomic</span></tt> Module</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.sequences.mixed"><tt class="docutils literal"><span class="pre">mixed</span></tt> Module</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.sequences.raw"><tt class="docutils literal"><span class="pre">raw</span></tt> Module</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.sequences.sequence"><tt class="docutils literal"><span class="pre">sequence</span></tt> Module</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="index.html#document-srcdocs/tigerlily.utility">utility Package</a><ul>
<li class="toctree-l5"><a class="reference internal" href="index.html#id1"><tt class="docutils literal"><span class="pre">utility</span></tt> Package</a></li>
<li class="toctree-l5"><a class="reference internal" href="index.html#module-tigerlily.utility.string_relations"><tt class="docutils literal"><span class="pre">string_relations</span></tt> Module</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li><a href="#">Tiger Lily 0.1-dev documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Erich Blume &lt;blume.erich@gmail.com&gt;.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1pre.
    </div>
  </body>
</html>